File: frontend/src/api/scores.ts
import client from './client'
import type { ScoreContent, ScoreListItem, ScoreRead } from '@/types/music'

export async function listScores(): Promise<ScoreListItem[]> {
  const res = await client.get('/scores')
  return res.data
}

export async function getScore(id: number): Promise<ScoreRead> {
  const res = await client.get(`/scores/${id}`)
  return res.data
}

export async function createScore(title: string, composer: string, content: ScoreContent): Promise<ScoreRead> {
  const res = await client.post('/scores', { title, composer, content })
  return res.data
}

export async function updateScore(id: number, data: { title?: string; composer?: string; content?: ScoreContent }): Promise<ScoreRead> {
  const res = await client.put(`/scores/${id}`, data)
  return res.data
}

export async function deleteScore(id: number): Promise<void> {
  await client.delete(`/scores/${id}`)
}

export async function validateScore(id: number): Promise<{ valid: boolean; errors: Record<number, string[]> }> {
  const res = await client.post(`/scores/${id}/validate`)
  return res.data
}

export function getMidiUrl(id: number, opts: { metronome?: boolean; start?: number; end?: number } = {}): string {
  const params = new URLSearchParams()
  if (opts.metronome) params.set('metronome', 'true')
  if (opts.start) params.set('start', String(opts.start))
  if (opts.end) params.set('end', String(opts.end))
  const qs = params.toString()
  return `/api/scores/${id}/midi${qs ? '?' + qs : ''}`
}

File: backend/app/routers/scores.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.music import ScoreContentSchema
from app.schemas.score import ScoreCreate, ScoreListItem, ScoreRead, ScoreUpdate
from app.services import score_service
from app.services.validation import validate_score
from app.services.score_service import schema_to_model

router = APIRouter(prefix="/api/scores", tags=["scores"])


@router.get("", response_model=list[ScoreListItem])
def list_scores(skip: int = 0, limit: int = 50, db: Session = Depends(get_db)):
    return score_service.list_scores(db, skip=skip, limit=limit)


@router.post("", response_model=ScoreRead, status_code=201)
def create_score(data: ScoreCreate, db: Session = Depends(get_db)):
    record = score_service.create_score(db, data)
    content = ScoreContentSchema.model_validate_json(record.content_json)
    return ScoreRead(
        id=record.id,
        title=record.title,
        composer=record.composer,
        created_at=record.created_at,
        updated_at=record.updated_at,
        content=content,
    )


@router.get("/{score_id}", response_model=ScoreRead)
def get_score(score_id: int, db: Session = Depends(get_db)):
    record = score_service.get_score(db, score_id)
    if not record:
        raise HTTPException(status_code=404, detail="Score not found")
    content = ScoreContentSchema.model_validate_json(record.content_json)
    return ScoreRead(
        id=record.id,
        title=record.title,
        composer=record.composer,
        created_at=record.created_at,
        updated_at=record.updated_at,
        content=content,
    )


@router.put("/{score_id}", response_model=ScoreRead)
def update_score(score_id: int, data: ScoreUpdate, db: Session = Depends(get_db)):
    record = score_service.update_score(db, score_id, data)
    if not record:
        raise HTTPException(status_code=404, detail="Score not found")
    content = ScoreContentSchema.model_validate_json(record.content_json)
    return ScoreRead(
        id=record.id,
        title=record.title,
        composer=record.composer,
        created_at=record.created_at,
        updated_at=record.updated_at,
        content=content,
    )


@router.delete("/{score_id}", status_code=204)
def delete_score(score_id: int, db: Session = Depends(get_db)):
    if not score_service.delete_score(db, score_id):
        raise HTTPException(status_code=404, detail="Score not found")


@router.post("/{score_id}/validate")
def validate(score_id: int, db: Session = Depends(get_db)):
    record = score_service.get_score(db, score_id)
    if not record:
        raise HTTPException(status_code=404, detail="Score not found")
    content = ScoreContentSchema.model_validate_json(record.content_json)
    score = schema_to_model(content)
    errors = validate_score(score)
    return {"valid": len(errors) == 0, "errors": errors}
